pyccel/codegen/printing/ccode.py

```diff
@@ -1033,15 +1036,18 @@ def _print_Constant(self, expr):
         return self._print(val)
 
     def _print_Return(self, expr):
-        code = ''
         args = [VariableAddress(a) if self.stored_in_c_pointer(a) else a for a in expr.expr]
-        if expr.stmt:
-            code += self._print(expr.stmt)+'\n'
+        if len(args) > 1:
+            if expr.stmt:
+                return self._print(expr.stmt)+'\n'+'return 0;'
+            return 'return 0;'
         if len(args) == 1:
-            code +='return {0};'.format(self._print(args[0]))
-        elif len(args) > 1:
-            code += 'return 0;'
-        return code
+            if isinstance(expr.stmt, CodeBlock):
+                if  isinstance(expr.stmt.body[-1], Deallocate):
+                    return '{0}\nreturn {1};'.format(self._print(expr.stmt), self._print(expr.stmt.body[-2].lhs))
```

Comment:
It is possible to have more than one `Deallocate` statement in a `Return`. For example, consider the following code:
```
def mat_mat_mat_mul(a : 'int[:,:]', b : 'int[:,:]', c : 'int[:,:]'):
    d = np.empty((a.shape[0], b.shape[1]))
    e = np.empty_like((b.shape[0], c.shape[1])
    d[:,:] = np.matmul(a,b)
    e[:,:] = np.matmul(b,c)
    a[:,:] = np.matmul(d,e)
    return True
```
This will be translated to:
```
FunctionDef(args = (Variable(a), Variable(b), Variable(c)),
                     result = Variable(Out_0001),
                     body = (
                           Allocate(d,...),
                           Allocate(e,...),
                           Assign(d[:,:], NumpyMatmul(a,b))
                           Assign(e[:,:], NumpyMatmul(b,c))
                           Assign(a[:,:], NumpyMatmul(d,e))
                           Return( expr = (Out_0001,)
                                        stmt = (
                                                Assign(Out_0001, LiteralTrue()),
                                                Deallocate(d),
                                                Deallocate(e)
                                               )
                                  )
                            )
           )
```

Reply:
I suggest you have a look at the construction of the `Return` node:
https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L2337-L2360

That way you should have a better idea of what it can contain and in what order